src/main/java/matteroverdrive/items/includes/EnergyContainer.java

```diff
@@ -25,28 +28,117 @@ public void setFull() {
         setEnergy(this.capacity);
     }
 
+    public EnergyContainer setItemStack(ItemStack stack) {
+        if (!stack.hasTagCompound()) {
+            NBTTagCompound tag = serializeNBT();
+            stack.setTagCompound(tag);
+        }
+
+        deserializeNBT(stack.getTagCompound());
+        this.stack = stack;
+        return this;
+    }
+
+    @Override
+    public int getEnergyStored() {
+        if (stack != null) {
+            return stack.getTagCompound().getInteger("mo_energystored");
+        }
+
+        return energy;
+    }
+
+    @Override
+    public int getMaxEnergyStored() {
+        if (stack != null) {
+            return stack.getTagCompound().getInteger("mo_capacity");
+        }
+
+        return capacity;
+    }
+
+    public int getMaxExtract() {
+        if (stack != null) {
+            return stack.getTagCompound().getInteger("mo_maxextract");
+        }
+
+        return maxExtract;
+    }
+
+    public int getMaxReceive() {
+        if (stack != null) {
+            return stack.getTagCompound().getInteger("mo_maxreceive");
+        }
+
+        return maxReceive;
+    }
+
     public void setEnergy(int energy) {
         this.energy = energy;
         if (this.energy > this.capacity)
             setFull();
     }
 
     @Override
+    public boolean canExtract() {
+        return getMaxExtract() > 0;
+    }
+
+    @Override
+    public boolean canReceive() {
+        return getMaxReceive() > 0;
+    }
+
+    @Override
+    public int receiveEnergy(int amount, boolean simulate) {
+        if (!canReceive())
+            return 0;
+
+        int energyReceived = Math.min(getMaxEnergyStored() - getEnergyStored(), Math.min(getMaxReceive(), amount));
+
+        if (!simulate) {
+            if (stack != null && energyReceived != 0) {
+                stack.getTagCompound().setInteger("mo_energystored", getEnergyStored() + energyReceived);
+            } else {
+                energy += energyReceived;
+            }
+        }
+
+        return energyReceived;
+    }
+
+    @Override
+    public int extractEnergy(int amount, boolean simulate) {
+        if (!canExtract())
+            return 0;
+
+        int energyExtracted = Math.min(getEnergyStored(), Math.min(getMaxExtract(), amount));
+
+        if (!simulate) {
+            if (stack != null && energyExtracted != 0) {
+                stack.getTagCompound().setInteger("mo_energystored", getEnergyStored() - energyExtracted);
+            } else {
+                energy -= energyExtracted;
+            }
+        }
+
+        return energyExtracted;
+    }
+
     public NBTTagCompound serializeNBT() {
         NBTTagCompound tag = new NBTTagCompound();
-        tag.setInteger("energy", energy);
-        tag.setInteger("capacity", capacity);
-        tag.setInteger("maxReceive", maxReceive);
-        tag.setInteger("maxExtract", maxExtract);
+        tag.setInteger("mo_energystored", energy);
```

Comment:
Is there a reason why all the nbt tags have to change? this breaks backward compat for no *real* reason

Reply:
It doesn't break. There is no backward compability, because before all tags were stored in Item/ForgeCaps/Parent, now they are in Item/nbt